fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/FDBStoreTimerTest.java

```diff
@@ -0,0 +1,192 @@
+        diffKVs = diffTimer.getKeysAndValues();
+        Assertions.assertTrue(diffKVs.containsKey("load_scan_entry_count"));
+        Assertions.assertEquals(diffKVs.get("load_scan_entry_count").intValue(), numAdvances);
+        Assertions.assertTrue(diffKVs.containsKey("load_key_value_count"));
+        Assertions.assertEquals(diffKVs.get("load_key_value_count").intValue(), numAdvances);
+    }
+
+    @Test
+    public void timeoutCounterDifferenceTest() {
+        RecordCursor<KeyValue> kvc = KeyValueCursor.Builder.withSubspace(subspace).setContext(context).setScanProperties(ScanProperties.FORWARD_SCAN).setRange(TupleRange.ALL).build();
+
+        FDBStoreTimer latestTimer = context.getTimer();
+        StoreTimerSnapshot savedTimer;
+        CompletableFuture<RecordCursorResult<KeyValue>> fkvr;
+        RecordCursorResult<KeyValue> kvr;
+        StoreTimer diffTimer;
+        Map<String, Number> diffKVs;
+
+        //generate a first timeout event advancing the cursor
+        context.getDatabase().setAsyncToSyncTimeout(1, TimeUnit.MILLISECONDS);
+        try {
+            context.getDatabase().asyncToSync(latestTimer, FDBStoreTimer.Waits.WAIT_ADVANCE_CURSOR, MoreAsyncUtil.delayedFuture(10000, TimeUnit.MILLISECONDS).thenCompose(ignore -> kvc.onNext()));
+        } catch (Exception e) {
+            Assertions.assertTrue(true);
+        }
+        //the latest timer should have recorded the one timeout event
+        diffKVs = latestTimer.getKeysAndValues();
+        Assertions.assertEquals(diffKVs.get("wait_advance_cursor_timeout_count").intValue(), 1);
+        Assertions.assertTrue(diffKVs.get("wait_advance_cursor_timeout_micros").intValue() > 0);
+
+        //record the state after the first timeout event and generate some more timeout events
+        savedTimer = StoreTimerSnapshot.from(latestTimer);
+        final int numTimeouts = 3;
+        for (int i = 0; i < numTimeouts; i++) {
+            try {
+                context.getDatabase().asyncToSync(latestTimer, FDBStoreTimer.Waits.WAIT_ADVANCE_CURSOR, MoreAsyncUtil.delayedFuture(10000, TimeUnit.MILLISECONDS).thenCompose(ignore -> kvc.onNext()));
+            } catch (Exception e) {
+                Assertions.assertTrue(true);
+            }
+        }
+        //should have the additional timeout events in latestTimer
+        diffKVs = latestTimer.getKeysAndValues();
+        Assertions.assertEquals(diffKVs.get("wait_advance_cursor_timeout_count").intValue(), numTimeouts + 1);
+        Assertions.assertTrue(diffKVs.get("wait_advance_cursor_timeout_micros").intValue() > 0);
+
+        //the savedTimer should only have recorded the first timeout event and hence the difference is the numTimeout events that occurred after that
+        diffTimer = StoreTimer.subtractSnapshot(latestTimer, savedTimer);
+        diffKVs = diffTimer.getKeysAndValues();
+        Assertions.assertEquals(diffKVs.get("wait_advance_cursor_timeout_count").intValue(), numTimeouts);
```

Comment:
I suspect that this test is passing because the `WAIT_ADVANCE_CURSOR` calls is equal to the number of `WAIT_ADVANCE_CURSOR` timeouts. If you added another one that didn't time out, I would think this test would fail due to the bug I think I identified above.

Reply:
This is true. Good catch. I changed the test case. I also reverted the code back to the way it was as this issues was introduced by changes made to modularize as per Scotts comments.